packages/pyright-internal/src/parser/tokenCollection.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ * tokenCollection.ts
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ *
+ * Stores a stream of tokens from a python file.
+ */
+
+import { assert } from '../common/debug';
+import { TextRangeCollection } from '../common/textRangeCollection';
+import { Token } from './tokenizerTypes';
+
+// Interface implemented by a TokenCollection
+interface ITokenCollection {
+    start: number;
+    end: number;
+    length: number;
+    count: number;
+    contains(position: number): boolean;
+    getItemAt(index: number): Token;
+    getItemAtPosition(position: number): number;
+}
+
+class TokenCollectionFast extends TextRangeCollection<Token> implements ITokenCollection {
+    constructor(private readonly _tokens: Token[]) {
+        super(_tokens);
+    }
+
+    toArray(): Token[] {
+        return this._tokens;
+    }
+}
+
+function* generateCompressedData(
+    tokens: Token[],
+    tokenPositions: Int32Array,
+    numberData: (number | bigint)[]
+): Generator<number, void, undefined> {
+    let currentPosition = 0;
+    for (let i = 0; i < tokens.length; i++) {
+        const token = tokens[i];
+        const tokenArray = Token.toCompressed(token, numberData);
+        tokenPositions[i] = currentPosition;
+        currentPosition += tokenArray.length;
+        yield* tokenArray;
+    }
+}
+
+// Special TokenCollection that is optimized for memory usage but is a lot
+// slower to fetch actual tokens.
+export class TokenCollectionCompressed implements ITokenCollection {
+    private _tokenData: Int32Array;
+    private _numberData: (number | bigint)[] = [];
```

Comment:
NumberTokens are special cased in that their data can be larger than 4GB. So I store the number value in a separate array.